File: enhanced-code/CS330/view_manager.py
# *****CODE INFORMATION*****
# Name: Michael King
# Date: 7/17/2026
# File: view_manager.py
# Assignment: CS-499 Enhancement 1
# CS-330 3D Scene Python Port
# University: SNHU

# Python Port of ViewManager.h/ViewManager.cpp from CS-330 (C++/OpenGL).
# ORIGINALLY MADE BY: Brian Battersby (SNHU Instructor).
# MODIFIED BY: Michael King
# Adapted for CS-499 Software Engineering Enhancement by Michael King
# PORTED TO PYTHON/PyOpenGL BY: Michael King

# The ViewManager class handles:
# - Building the Camera instance with the scene's initial position/orientation
# - Processing keyboard input each frame (WASD + Q/E + P/O + ESC)
# - Processing mouse-look input (cursor delta -> yaw/pitch)
# - Processing scroll-wheel input (movement speed adjustment)
# - Building the view and projection matrices each frame
# - Uploading view, projection and viewPosition uniforms to the shader

# Key Differences from C++:
# - GLFW callbacks are replaced with Pygame event polling. Pygame is the
#   windowing system used by this Python port in place of GLFW/GLEW)
# - Static callback pattern is replaced with instance methods
# - glm::mat4 is replaced with pyrr matrix functions and numpy arrays

# CONTROLS (IDENTICAL TO THE C++ VERSION):  
# W / S...................Move Forward/Backward
# A / D...................Pan Left/Right
# Q / E...................Move Up/Down
# Mouse...................Look Around (Yaw/Pitch)
# Scroll Wheel............Adjust Camera Speed
# P.......................Perspective (3D) projection
# O.......................Orthographic (2D) projection
# ESC.....................Quit

# *** Python Imports ***
import numpy as np
import pyrr
import pygame

from shader_manager import ShaderManager
from camera import (Camera, FORWARD, BACKWARD, LEFT, RIGHT, UP, DOWN)

# *****************************************************************************
# WINDOW DIMENSIONS; match ViewManager.cpp namespace constants
# *****************************************************************************
WINDOW_WIDTH = 800
WINDOW_HEIGHT = 600

# *****************************************************************************
# SHADER UNIFORM NAMES; match the GLSL uniform declarations
# *****************************************************************************
U_VIEW       = "view"
U_PROJECTION = "projection"
U_VIEW_POS   = "viewPosition"

# *****************************************************************************
# Default camera movement speed and scroll step
# From ViewManager.cpp
# *****************************************************************************
DEFAULT_SPEED = 2.5  #units per second
SPEED_STEP    = 0.5  #units per second per scroll tick
SPEED_MIN     = 0.5  #minimum scroll speed
SPEED_MAX     = 20.0 #maximum scroll speed

class ViewManager:
    '''
    Manages the 3D view: camera, input processing, and matrix uploads.

    Mirrors ViewManager.h / ViewManager.cpp from CS-330.
    Replaces GLFW callbacks with Pygame event handling.
    '''

    def __init__(self, shader_manager: ShaderManager):
        '''
        Initialize the view manager and the camera.

        Mirrors ViewManager::ViewManager() from ViewManager.cpp
        Sets the camera's initial position, orientation, FOV, and
        speed to the same values use in the C++ version.

        ARGS:
        shader_manager: The active ShaderManager.
                        Receives view and projection matrix uploads
                        each frame.
        '''
        self._shader = shader_manager

        # Instantiate the camera
        # Mirrors "g_pCamera = new Camera()" in C++
        self._camera = Camera()
    

        #--- Initial camera values from ViewManager.cpp ---
        #Place the camera above and behind the scene center so that all
        #objects are visible when the application launches
        self._camera.position = np.array([0.0, 5.0, 12.0], dtype=np.float32)

        #Aim slightly downward so that the scene appears centered on screen
        #Front is normalized internally by the Camera class
        self._camera.front = np.array([0.0, -0.5, -2.0], dtype=np.float32)
        self._camera.front /= np.linalg.norm(self._camera.front)

        #World-up vector keeps the camera from rolling sideways
        self._camera.world_up = np.array([0.0, 1.0, 0.0], dtype=np.float32)

        #80-degree FOV matches g_pCamera->Zoom = 80 in C++ constructor
        self._camera.zoom = 80.0

        #Movement speed; matches the default gMovementSpeed
        self._camera.movement_speed = DEFAULT_SPEED

        # *** Mouse Tracking State ***
        # Mirrors gLastX, gLastY, gFirstMouse
        self._last_x      = WINDOW_WIDTH  / 2.0
        self._last_y      = WINDOW_HEIGHT / 2.0
        self._first_mouse = True # Prevents a jump on the first move

        # *** Frame Timing ***
        # Mirrors gDeltaTime / gLastFrame
        self._last_frame_ms = pygame.time.get_ticks()

        # *** Projection Mode Toggle ***
        # Mirrors bOrthographicProjection
        self._orthographic = False # False = perspective (3D); True = orthographic (2D)
        
        # *** Key State Tracker for Toggle Debouncing ***
        # P / O keys
        self._prev_keys = pygame.key.get_pressed()

    # *****************************************************************************
    # Public API
    # *****************************************************************************

    @property
    def camera(self) -> Camera:
        '''expose the camera for external read (e.g., main loop HUD)'''
        return self._camera

    @property
    def should_quit(self) -> bool:
        '''true if the user pressed ESC or closed the window'''
        return self._quit_flag
    
    def prepare_scene_view(self) -> bool:
        '''
        Compute delta time, process input, build matrices, upload to shader.

        Mirrors ViewManager::PrepareSceneView() from ViewManager.cpp
        Must be called once per frame before drawing.

        Returns true if the application should continue; 
        false if ESC was pressed.
        '''
        # *** DELTA TIME; mirrors gDeltaTime computation ***
        now_ms     = pygame.time.get_ticks()
        delta_time = (now_ms - self._last_frame_ms) / 1000.0 # seconds
        self._last_frame_ms = now_ms

        # *** Poll Keyboard and Mouse ***
        if not self._process_keyboard(delta_time):
            return False  # ESC Pressed; signal the main loop to exit
        self._process_mouse_events()

        # *** Build View Matrix from Current Camera State ***
        view = self._camera.get_view_matrix()

        # *** Build projection matrix (Perspective or Orthographic) ***
        if self._orthographic:
            #Orthographic: scale=0.01 converts world units to NDC
            #Matches glm::ortho() call in ViewManager.cpp
            scale = 0.01
            projection = pyrr.matrix44.create_orthogonal_projection(
                left   = -WINDOW_WIDTH  * scale,
                right  =  WINDOW_WIDTH  * scale,
                bottom = -WINDOW_HEIGHT * scale,
                top    =  WINDOW_HEIGHT * scale,
                near   = 0.1,
                far    = 100.0,
                dtype  = np.float32,
            )
        else:
            #Perspective: standard 3D projection
            projection = pyrr.matrix44.create_perspective_projection(
                fovy = self._camera.zoom,
                aspect = WINDOW_WIDTH / WINDOW_HEIGHT,
                near = 0.1,
                far = 100.0,
                dtype = np.float32,
            )
        
        # *** Upload Matrices to the Shader ***
        self._shader.set_mat4("view",       view)
        self._shader.set_mat4("projection", projection)
        self._shader.set_vec3("viewPosition",
                              *self._camera.position.tolist())
        
        return True
    
    def handle_event(self, event: pygame.event.Event):
        '''
        Handle a single Pygame event.

        Called by the main loop for events that must be processed outside of the 
        per-frame keyboard poll, specifically scroll-wheel events.

        Mirrors Mouse_Scroll_Callback() in ViewManager.cpp

        ARGS:
        event: A Pygame event object.
        '''
        if event.type == pygame.MOUSEWHEEL:       
            #SCROLL UP (y > O): increase speed
            #SCROLL DOWN (y < 0): decrease speed
            #Multiply by SPEED_STEP (0.5) per tick; matches C++ yOffset * 0.5
            new_speed = self._camera.movement_speed + event.y * SPEED_STEP
            new_speed = max(SPEED_MIN, min(SPEED_MAX, new_speed))
            if new_speed != self._camera.movement_speed:
                self._camera.movement_speed = new_speed
                print(f"Camera movement speed: {new_speed:.1f}")

    # *****************************************************************************
    # PRIVATE INPUT HELPERS
    # *****************************************************************************
    def _process_keyboard(self, delta_time: float) -> bool:
        '''
        Poll the keyboard state and apply camera movements.

        Mirrors ViewManager::ProcessKeyboardEvents() from ViewManager.cpp.
        Multiple keys can be held simultaneously (e.g. W+D = move diagonally).

        ARGS:
        delta_time: Elapsed seconds since last frame for speed scaling.

        Returns False if ESC was pressed (signals the main loop to quit), 
                True otherwise.
        '''
        keys = pygame.key.get_pressed()

        #ESC: Close the Application
        if keys[pygame.K_ESCAPE]:
            return False
        
        # W / S: Forward / Backward
        if keys[pygame.K_w]:
            self._camera.process_keyboard(FORWARD, delta_time)
        if keys[pygame.K_s]:
            self._camera.process_keyboard(BACKWARD, delta_time)

        # A / D: Strafe Left / Right
        if keys[pygame.K_a]:
            self._camera.process_keyboard(LEFT, delta_time)
        if keys[pygame.K_d]:
            self._camera.process_keyboard(RIGHT, delta_time)

        # Q / E: Move Up / Down
        if keys[pygame.K_q]:
            self._camera.process_keyboard(UP, delta_time)
        if keys[pygame.K_e]:
            self._camera.process_keyboard(DOWN, delta_time)

        # P: Switch to Perspective (3D) projection; debounced
        if keys[pygame.K_p] and not self._prev_keys[pygame.K_p]:
            self._orthographic = False
            print("Projection: PERSPECTIVE")

        # O: Switch to Orthographic (2D) projection; debounced
        if keys[pygame.K_o] and not self._prev_keys[pygame.K_o]:
            self._orthographic = True
            print("Projection: ORTHOGRAPHIC")
        
        self._prev_keys = keys
        return True
    
    def _process_mouse_events(self):
        '''
        Compute cursor delta from the current relative mouse motion
        and forward it to the camera for look-around.

        Mirrors Mouse_Position_Callback() in ViewManager.cpp.
        In Pygame, pygame.mouse.get.rel() returns the pixel delta since the
        last call; equivalent to the (xMousePos - gLastX) calculation in C++.
        The mouse is captured with MOUSEMOTION in relative mode.
        '''
        dx, dy = pygame.mouse.get_rel()
        if dx == 0 and dy == 0:
            return
        
        # Invert Y so that moving the mouse up pitches the camera upward,
        # matching the "yOffset = gLastY - yMousePos" inversion in C++.
        self._camera.process_mouse_movement(float(dx), float(-dy))

File: original-code/CS330/CS330Content/Projects/7-1_FinalProjectMilestones/Source/ViewManager.cpp
///////////////////////////////////////////////////////////////////////////////
// viewmanager.h
// ============
// manage the viewing of 3D objects within the viewport
//
//  AUTHOR: Brian Battersby - SNHU Instructor / Computer Science
//	Created for CS-330-Computational Graphics and Visualization, Nov. 1st, 2023
// 
//  MODIFIED BY: Michael King
//  ********************************************
//  *            MILESTONE THREE               *
//  ********************************************
//  CHANGES MADE:
//  - Implemented Mouse_Position_Callback() for look-around (pitch/yaw)
//  - Implemented scroll callback for camera speed adjustment
//  - Implemented WASD + QE keyboard camera navigation
//  - Enabled cursor capture for smooth mouse look
// 
//  *********************************************
//  *             FINAL PROJECT                 *
//  *********************************************
//  CHANGES MADE:
//  - Added P / O key toggle between perspective and orthographic projection
//    P key = Perspective (3D)
//    O key = Orthographic (2D)
//  - bOrthographicProjection flag switches which glm matrix is built
//    in PrepareSceneView() and uploaded to the shader each frame.
//  - Camera stays in the same orientation when switching projections.
// 
// CONTROLS:
// - W / S.................Forward (zoom in) / Backward (zoom out)
// - A / D.................Pan Left/Right
// - Q / E.................Move Up/Down
// - Mouse.................Look Around (Yaw / Pitch)
// - Scroll Up.............Increase Camera Movement Speed
// - Scroll Down...........Decrease Camera Movement Speed
// - P.....................Switch to Perspective (3D) view
// - O.....................Switch to Orthographic (2D) view
// - Esc...................Close Application
///////////////////////////////////////////////////////////////////////////////

#include "ViewManager.h"

// GLM Math Header inclusions
#include <glm/glm.hpp>
#include <glm/gtx/transform.hpp>
#include <glm/gtc/type_ptr.hpp>    

// declaration of the global variables and defines
namespace
{
	// Variables for window width and height
	const int WINDOW_WIDTH = 1000;
	const int WINDOW_HEIGHT = 800;

	//Uniform names must match the variable names declared in vertex shader
	const char* g_ViewName = "view";
	const char* g_ProjectionName = "projection";

	// camera object used for viewing and interacting with
	// the 3D scene
	Camera* g_pCamera = nullptr;

	// these variables are used for mouse movement processing
	float gLastX = WINDOW_WIDTH / 2.0f;
	float gLastY = WINDOW_HEIGHT / 2.0f;
	bool gFirstMouse = true;

	// gDeltaTime is the elapsed time (in Seconds) between the current and previous frame.
	// Multiplying movement amounts by gDeltaTime keeps camera speed consistent regardless
	// of how fast or slow the machine is rendering frames.
	float gDeltaTime = 0.0f;
	float gLastFrame = 0.0f;

	//Camera movement speed is adjusted by mouse scroll wheel
	//The default is 2.5 units/sec
	//Scroll up increases
	//Scroll down decreases
	//Clamped to [0.5, 20.0] so it remains usable
	float gMovementSpeed = 2.5f;

	//This variable is false when orthographic projection is off
	//and true when it is on
	bool bOrthographicProjection = false;
}

/***********************************************************
 *  ViewManager()
 *
 *  The constructor for the class
 ***********************************************************/
ViewManager::ViewManager(
	ShaderManager *pShaderManager)
{
	//Store the ShaderManager pointer so PrepareSceneView() can
	//upload the view and projection matrices to the GPU each frame.
	m_pShaderManager = pShaderManager;
	m_pWindow = NULL;
	g_pCamera = new Camera(); //allocate the camera object on the heap.
	
	// default camera view parameters
	// Place the camera above and behind the origin
	// so all scene objects are visible when app launches
	g_pCamera->Position = glm::vec3(0.0f, 5.0f, 12.0f);

	//Aim slightly downward towards the scene center so it
	//appears centered on the screen.
	g_pCamera->Front = glm::vec3(0.0f, -0.5f, -2.0f);

	//World up vector
	//keeps the camera from rolling sideways
	g_pCamera->Up = glm::vec3(0.0f, 1.0f, 0.0f);

	//Field-of-view angle in degrees
	//Used when building the projection matrix
	g_pCamera->Zoom = 80;

	//Apply Initial Movement Speed to the camera
	g_pCamera->MovementSpeed = gMovementSpeed;
}

/***********************************************************
 *  ~ViewManager()
 *
 *  The destructor for the class
 ***********************************************************/
ViewManager::~ViewManager()
{
	// Null out the borrowed pointer
	// This class does not own the Shader Manager.
	m_pShaderManager = NULL;
	m_pWindow = NULL;

	//Free the camera that was allocated in the constructor.
	if (NULL != g_pCamera)
	{
		delete g_pCamera;
		g_pCamera = NULL;
	}
}

/***********************************************************
 *  CreateDisplayWindow()
 *
 *  Creates the GLFW window and OpenGL context.
 *  Registers all three input callbacks:
 *  - Cursor
 *  - Scroll
 *  - Keyboard (glfwGetKey)
 *  Returns the window handle to the caller (main.cpp)
 ***********************************************************/
GLFWwindow* ViewManager::CreateDisplayWindow(const char* windowTitle)
{
	GLFWwindow* window = nullptr;

	// try to create the displayed OpenGL window
	// pass NULL for monitor to keep it windowed, not Fullscreen.
	window = glfwCreateWindow(
		WINDOW_WIDTH,
		WINDOW_HEIGHT,
		windowTitle,
		NULL, NULL);
	if (window == NULL)
	{
		//Window creation fails if the requested OpenGL version is unsupported.
		std::cout << "Failed to create GLFW window" << std::endl;
		glfwTerminate();
		return NULL;
	}

	//Make this window's OpenGL context active on the calling thread
	glfwMakeContextCurrent(window);

	// Register the cursor-movement callback
	// GLFW invokes this function automatically whenever the mouse
	// moves inside the window.
	// It feeds the new cursor coordinates to Mouse_Position_Callback(),
	// which computes the delta and calls g_pCamera->ProcessMouseMovement().
	glfwSetCursorPosCallback(window, &ViewManager::Mouse_Position_Callback);

	// Register the scroll-wheel callback.
	// GLFW invokes this function automatically on every scroll event.
	// It adjusts gMovement speed and pushes the new value into the camera object.
	glfwSetScrollCallback(window, &ViewManager::Mouse_Scroll_Callback);

	// Hide and confine the cursor to the window for FPS-style look around.
	// Without GLFW_CURSOR_DISABLED, the cursor stays visible and can leave the window,
	// which can break the mouse-delta calculation in the callback.
	glfwSetInputMode(window, GLFW_CURSOR, GLFW_CURSOR_DISABLED);

	// enable blending for supporting transparent rendering
	glEnable(GL_BLEND);
	glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);

	m_pWindow = window;

	return(window);
}

/***********************************************************
 *  Mouse_Position_Callback()
 *
 *  This method is automatically called from GLFW whenever
 *  the mouse is moved within the active GLFW display window.
 * 
 *  It calculates how far the cursor moved
 *  (xOffset, yOffset)
 * 
 *	Passes those offsets to the camera so it can update the
 *  yaw (left / right) and
 *  pitch (up / down) angles.
 ***********************************************************/
void ViewManager::Mouse_Position_Callback(GLFWwindow* window, double xMousePos, double yMousePos)
{
	// when the first mouse move event is received, this needs to be recorded so that
	// all subsequent mouse moves can correctly calculate the
	// X position offset and Y position offset for proper operation

	if (gFirstMouse) {
		gLastX = xMousePos;
		gLastY = yMousePos;
		gFirstMouse = false;
	}

	//calculate the X offset and the Y offset values for moving the 3D camera accordingly
	float xOffset = xMousePos - gLastX;
	float yOffset = gLastY - yMousePos; //reversed since the y-coordinates go from bottom to top

	//set the current positions into the last position variables
	gLastX = xMousePos;
	gLastY = yMousePos;

	// Move the 3D Camera according to the calculated offsets
	g_pCamera->ProcessMouseMovement(xOffset, yOffset);
}

/***********************************************************
 * Mouse_Scroll_Callback()
 *
 * Called by GLFW on every scroll-wheel event.
 * Increases or decreases the camera movement speed so the
 * user can navigate large or tight scenes.
 *
 * --- Method Controls --
 * SCROLL UP.........Increase Camera Movement Speed
 *                   +0.5/tick
 * SCROLL DOWN.......Decrease Camera Movement Speed
 *                   -0.5/tick
 * SPEED.............[0.5, 20.0] to stay usable and stable.
 * 
 * --- Parameters ---
 * window............GLFW window that received the event.
 * xOffset...........Horizontal scroll amount (not used here)
 * yOffset...........Vertical scroll
 *                   +1.0/tick up; -1.0/tick down
 ************************************************************/

void ViewManager::Mouse_Scroll_Callback(GLFWwindow* window, double xOffset, double yOffset) {
	
	// Scroll up = yOffset is positive
	// Scroll down = yOffset is negative
	// Multiply by 0.5 so each scroll tick makes a small change
	gMovementSpeed += static_cast<float>(yOffset) * 0.5f;

	//Lower bound: prevent the speed from reaching 0 or below,
	//which would freeze or reverse all keyboard movement
	if (gMovementSpeed < 0.5f) gMovementSpeed = 0.5f;

	//Upper bound: prevent the speed from becoming so large that a
	//single frame moves the camera entirely across the scene.
	if (gMovementSpeed > 20.0f) gMovementSpeed = 20.0f;

	//Push the updated speed into the camera so ProcessKeyboard()
	//picks it up on the very next frame
	if (g_pCamera != NULL) {
		g_pCamera->MovementSpeed = gMovementSpeed;
	}

	//Display current speed on the console for verification/debugging
	std::cout << "Camera movement speed: " << gMovementSpeed << std::endl;
}


/***********************************************************
 *  ProcessKeyboardEvents()
 *
 *  Polls the keyboard state once per frame and issues the
 *  appropriate movement command to the camera.
 *  Multiple keys may be held simultaneously for combined
 *  movement.
 * 
 * Example:
 * W + D..........Move forward and right at the same time.
 * 
 * All movement is scaled by gDeltaTime
 * (computed in PrepareSceneView)
 * so speed stays consistent across different frame rates.
 * 
 * CONTROLS:
 * W / S..........Move Forward  (Zoom In)
 *                Move Backward (Zoom Out)
 * A / D..........Pan Left
 *                Pan Right (Strafe)
 * Q / E..........Move Up
 *                Move Down
 * ESC............Close Application
 ***********************************************************/
void ViewManager::ProcessKeyboardEvents()
{
	// signal GLFW to close the window if ESC key is pressed
	if (glfwGetKey(m_pWindow, GLFW_KEY_ESCAPE) == GLFW_PRESS)
	{
		glfwSetWindowShouldClose(m_pWindow, true);
	}

	//Nothing below can execute safely without a valid camera pointer
	if (NULL == g_pCamera) { return; }

	// ---- W / S: Forward and Backward (Zooming In and Out) ----
	// The camera moves along its front vector so it always advances or
	// retreats in exactly the direction the user is looking
	if (glfwGetKey(m_pWindow, GLFW_KEY_W) == GLFW_PRESS)
	{
		g_pCamera->ProcessKeyboard(FORWARD, gDeltaTime); //PRESS W
	}if (glfwGetKey(m_pWindow, GLFW_KEY_S) == GLFW_PRESS)
	{
		g_pCamera->ProcessKeyboard(BACKWARD, gDeltaTime); //PRESS S
	}

	// ---- A / D: Pan Left and Right (Strafe) ----
	// The camera moves along its right vector, which is perpendicular to the front.
	// This slides the camera horizontally without changing its aim.
	if (glfwGetKey(m_pWindow, GLFW_KEY_A) == GLFW_PRESS) {
		g_pCamera->ProcessKeyboard(LEFT, gDeltaTime); //PRESS A
	} if (glfwGetKey(m_pWindow, GLFW_KEY_D) == GLFW_PRESS) {
		g_pCamera->ProcessKeyboard(RIGHT, gDeltaTime); //PRESS D
	}

	// ---- Q / E: Move Up and Down (Vertical) ----
	// The camera moves along the up vector, rising/sinking vertically,
	// regardless of which horizontal direction it is facing.
	if (glfwGetKey(m_pWindow, GLFW_KEY_Q) == GLFW_PRESS) {
		g_pCamera->ProcessKeyboard(UP, gDeltaTime); //PRESS Q
	} if (glfwGetKey(m_pWindow, GLFW_KEY_E) == GLFW_PRESS) {
		g_pCamera->ProcessKeyboard(DOWN, gDeltaTime); //PRESS E
	}

	// ---- P: Switch to Perspective (3D) Projection ----
	// Pressing P restores the standard 3D perspective view.
	// The camera stays in its current position and orientation.
	if (glfwGetKey(m_pWindow, GLFW_KEY_P) == GLFW_PRESS) {
		bOrthographicProjection = false;
	}

	// ---- O: Switch to Orthographic (2D) Projection ----
	// Pressing O flattens the view to a 2D orthographic projection.
	// The camera stays in its current position and orientation.
	if (glfwGetKey(m_pWindow, GLFW_KEY_O) == GLFW_PRESS) {
		bOrthographicProjection = true;
	}
}

/***********************************************************
 *  PrepareSceneView()
 *
 *  This method is used for preparing the 3D scene by loading
 *  the shapes, textures in memory to support the 3D scene 
 *  rendering
 * 
 * Responsibilities:
 * 1. Compute gDeltaTime for frame-rate-independent movement.
 * 2. Poll keyboard input via ProcessKeyboardEvents().
 * 3. Retrieve the current view matrix from the camera.
 * 4. Build the projection matrix: either perspective (3D)
 *    or orthographic (2D) depending on bOrthographicProjection.
 * 5. Upload view, projection, and camera position to the shader.
 * 
 * The P / O keys in ProcessKeyboardEvents() toggle
 * bOrthographicProjection, causing this function to switch
 * which matrix is sent to the GPU on the next frame.
 ***********************************************************/
void ViewManager::PrepareSceneView()
{
	glm::mat4 view;
	glm::mat4 projection;

	// per-frame timing
	float currentFrame = glfwGetTime();
	gDeltaTime = currentFrame - gLastFrame;
	gLastFrame = currentFrame;

	// process any keyboard events that may be waiting in the 
	// event queue
	ProcessKeyboardEvents();

	// get the current view matrix from the camera
	view = g_pCamera->GetViewMatrix();

	//Build the projection matrix based on the current mode.
	if (bOrthographicProjection) {

		// ORTHOGRAPHIC (2D) PROJECTION
		// Projects the scene without perspective foreshortening.
		// Objects appear the same size regardless of their distance.
		// The visible volume is defined by left/right/bottom/top/near/far.
		// Scale factor of 0.01 coverts from pixel-like coordinates to
		// world units
		float scale = 0.01f;
		projection = glm::ortho(
			-(float)WINDOW_WIDTH * scale,  //left
			(float)WINDOW_WIDTH * scale,   //right
			-(float)WINDOW_HEIGHT * scale, //bottom
			(float)WINDOW_HEIGHT * scale,  //top
			0.1f,                          //near clip
			100.0f);                       //far clip
	}
	else {
		// PERSPECTIVE (3D) PROJECTION
		// Projects the scene with perspective foreshadowing so distant
		// objects appear smaller; the standard 3D view.
		projection = glm::perspective(
			glm::radians(g_pCamera->Zoom),                  //Field of view
			(GLfloat)WINDOW_WIDTH / (GLfloat)WINDOW_HEIGHT, //Aspect Ratio
			0.1f,                                           //near clip pane
			100.0f);                                        //far clip pane
	}

	// if the shader manager object is valid
	if (NULL != m_pShaderManager)
	{
		// set the view matrix into the shader for proper rendering
		m_pShaderManager->setMat4Value(g_ViewName, view);
		// set the view matrix into the shader for proper rendering
		m_pShaderManager->setMat4Value(g_ProjectionName, projection);
		// set the view position of the camera into the shader for proper rendering
		m_pShaderManager->setVec3Value("viewPosition", g_pCamera->Position);
	}
}